[te] Fix clamp with uint8 args

bertmaher opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #49271 [te] Fix bugs with shift operators
* #49247 [te] Ban uint8 tensors from fusion groups
* #49148 [te] Use c10::ScalarType utility functions in te::Dtype
* #49147 [te] Use Dtype::is_signed instead of an ad hoc local predicate.
* **#49143 [te] Fix clamp with uint8 args**

Riddle me this, batman: how could `torch.clamp(torch.tensor([0], dtype=torch.uint8), -10, 10)` equal `10`?  The answer: the min/max args are first cast to the dtype of the input, giving min=246 and max 10.  Then you have to apply Min and Max in the right order: `Min(Max(in, min), max)`.  Differ in any way and you're doomed.  Hooray.

This PR makes TE match eager mode for this operator, plus fixes a major facepalm in the llvm min/max codegen where we were always generating signed comparisons.

Differential Revision: [D25456366](https://our.internmc.facebook.com/intern/diff/D25456366/)

bertmaher committed:
[te] Fix clamp with uint8 args

Riddle me this, batman: how could `torch.clamp(torch.tensor([0], dtype=torch.uint8), -10, 10)` equal `10`?  The answer: the min/max args are first cast to the dtype of the input, giving min=246 and max 10.  Then you have to apply Min and Max in the right order: `Min(Max(in, min), max)`.  Differ in any way and you're doomed.  Hooray.

This PR makes TE match eager mode for this operator, plus fixes a major facepalm in the llvm min/max codegen where we were always generating signed comparisons.

Differential Revision: [D25456366](https://our.internmc.facebook.com/intern/diff/D25456366/)

[ghstack-poisoned]

bertmaher committed:
clang-format on "[te] Fix clamp with uint8 args"


Riddle me this, batman: how could `torch.clamp(torch.tensor([0], dtype=torch.uint8), -10, 10)` equal `10`?  The answer: the min/max args are first cast to the dtype of the input, giving min=246 and max 10.  Then you have to apply Min and Max in the right order: `Min(Max(in, min), max)`.  Differ in any way and you're doomed.  Hooray.

This PR makes TE match eager mode for this operator, plus fixes a major facepalm in the llvm min/max codegen where we were always generating signed comparisons.

Differential Revision: [D25456366](https://our.internmc.facebook.com/intern/diff/D25456366/)

[ghstack-poisoned]

zheng-xq commented on torch/csrc/jit/tensorexpr/kernel.h:
```diff
@@ -101,7 +101,8 @@ class TORCH_API TensorExprKernel {
           ExprHandle(const ExprHandle&, const ExprHandle&, const ExprHandle&)>&
-          innerExpr);
+          innerExpr,
+      bool promote_inputs = true);
```
Minor: it doesn't feel that "promote_inputs" is true/false is two flavors of this function. My first impression is that "promote_inputs = true" is the public behavior, and "promote_inputs = false" is an internal behavior. I would prefer to remove this argument from the public function, and refactor the "= false" flavor to a "_interna"/"_impl" function.

zheng-xq commented on torch/csrc/jit/tensorexpr/types.cpp:
```diff
@@ -9,6 +9,10 @@ namespace torch {
 namespace tensorexpr {
 
+static bool is_c10_type(const ScalarType& type) {
+  return type < ScalarType::Undefined;
```
Minor: this is fine. But it feels a bit dangerous to rely on enum value ordering. I've seen in other systems when a list of enums is imported, it also uses the same macros to define a list of type traits on those enums. Otherwise, the enum value might keep changing, which makes serialization compatibility a bit complicated. It is up to you.

bertmaher replied:
Right, so the main reason I want to use ordering here is that we define the first N values of te::ScalarType in lockstep with c10::ScalarType (and we depend on that all over the place).  Maybe something less brittle would be better.  But I think I'd rather put the effort towards making te::ScalarType just go away in favor of c10::ScalarType, if possible. :)

asuhan commented on torch/csrc/jit/tensorexpr/types.cpp:
```diff
@@ -38,6 +42,13 @@ bool is_floating_point(const ScalarType& type) {
   return false;
 }
 
+bool is_signed(const ScalarType& type) {
```
The generality of `is_signed` confused me a bit - given that it's only used on paths handling integrals anyway, I don't know if it's warranted. If we made it handle integrals only, it'd side-step the `is_c10_type` trick as well.

asuhan commented on torch/csrc/jit/tensorexpr/llvm_codegen.cpp:
```diff
@@ -720,7 +720,8 @@ void LLVMCodeGenImpl::visit(const Max* v) {
 
   if (v->dtype().is_integral()) {
-    auto icmp = irb_.CreateICmpSGT(lhs, rhs);
+    auto icmp = v->dtype().is_signed() ? irb_.CreateICmpSGT(lhs, rhs)
```
You could use `CreateICmp` and `llvm_comparison_predicate` here and avoid the ternary - I had to do a similar fix for comparisons.

ZolotukhinM commented on torch/csrc/jit/tensorexpr/kernel.cpp:
```diff
@@ -688,7 +689,9 @@ Tensor* TensorExprKernel::computeThreeOperand(
         };
 
-        promoteInputs(inputs);
+        if (promote_inputs) {
```
Should we still be demoting output with this flag?

bertmaher replied:
Yes, if I understand correctly the output type should still be unchanged.

eellison commented:
Related: https://github.com/pytorch/pytorch/issues/49178

bertmaher committed:
gpu fix and rebase on "[te] Fix clamp with uint8 args"


Riddle me this, batman: how could `torch.clamp(torch.tensor([0], dtype=torch.uint8), -10, 10)` equal `10`?  The answer: the min/max args are first cast to the dtype of the input, giving min=246 and max 10.  Then you have to apply Min and Max in the right order: `Min(Max(in, min), max)`.  Differ in any way and you're doomed.  Hooray.

This PR makes TE match eager mode for this operator, plus fixes a major facepalm in the llvm min/max codegen where we were always generating signed comparisons.

Differential Revision: [D25456366](https://our.internmc.facebook.com/intern/diff/D25456366/)

[ghstack-poisoned]

bertmaher committed:
fix test on "[te] Fix clamp with uint8 args"


Riddle me this, batman: how could `torch.clamp(torch.tensor([0], dtype=torch.uint8), -10, 10)` equal `10`?  The answer: the min/max args are first cast to the dtype of the input, giving min=246 and max 10.  Then you have to apply Min and Max in the right order: `Min(Max(in, min), max)`.  Differ in any way and you're doomed.  Hooray.

This PR makes TE match eager mode for this operator, plus fixes a major facepalm in the llvm min/max codegen where we were always generating signed comparisons.

Differential Revision: [D25456366](https://our.internmc.facebook.com/intern/diff/D25456366/)

[ghstack-poisoned]
